prime-router/docs/design/design/auth/auth-design.md

```diff
@@ -0,0 +1,431 @@
+
+## Requirements
+
+- Protect certain endpoints from only being accessed by identified trusted entities
+  - Reports/Waters report submission endpoint
+  - History endpoints
+  - Delivery endpoints
+  - Message endpoints
+  - Admin endpoints for CRUD operations on orgs, senders, and receivers
+  - Lookup table CRUD operation endpoints
+- Centralize user management
+- Revoke access if a secret is compromised
+- Easily audit which clients are accessing which resources
+- Fine-tune access control for specific clients
+- Unify authn/authz approach for all microservices
+- Allow clients the ability to self-service to ease client onboarding
+  - [Okta documentation for this workflow](https://help.okta.com/en-us/content/topics/apps/apps-selfservice-configureorg.htm)
+
+## Options discussed in 14765
+
+- Azure Application Gateway
+  - This is Azure's load balancer and does not handle authn/authz
+- Azure Front Door
+  - This is Azure's CDN technology and does not handle authn/authz
+- Azure Identity Management
+  - This is Azure's own implementation of handling authn/authz for Microsoft products and is not available for implementation in our application
+
+## Identity Provider Options
+
+### Azure AD
+
+Pros
+- Integrated into the Microsoft ecosystem that we currently reside
+- Scales well
+
+Cons
+- Complex setup
+- Requires coordination with the CDC
+ 
+### Okta
+
+Pros
+- A trusted industry standard
+- Highly scalable
+- Offers self-service to ease Engagement teams workload
+- We already have many of our clients set up here!
+
+Cons
+- Single point of failure if there is an outage
+
+### ReportStream implementation
+
+Pros
+- As flexible as we need it to be
+- Cheap
+
+Cons
+- Custom authn/authz code is often vulnerable to attacks we may not have considered
+- Additional work for us to write and maintain
+
+## Protocol
+
+We will be using OAuth 2.0 with OIDC. OAuth 2.0 is an industry-standard authorization protocol and OIDC is
+another open standard on top of OAuth that adds an identity layer (authentication).
+
+We will be using the **client credentials** flow which was designed with machine-to-machine communication in 
+mind. See the image below for how the flow works generally.
+
+![Client Credentials Flow](clientCredentialsFlow.png)
+
+Users with emails and passwords will use the Interaction Code flow. Most of the interaction occurs by browser redirects between
+the frontend and authorization server. From the platform perspective it will be handled similarly once the access token
+is retrieved.
+
+## Definitions
+
+- **authn**: Authenticate/authentication
+- **authz**: Authorize/authorization
+- **IAM**: Identity and access management
+- **IdP**: Identity Provider
+- **Okta**: A cloud-hosted identity and access management service
+- **Server2Server**: ReportStream's implementation JWT authn/authz
+- **VPC**: Virtual Private Cloud
+- **Resource Server**: A server that will require authn/authz to access. In our case ResportStream
```

Comment:
CTRL-F for `ResportStream` and fix spelling, occurs multiple times.

Reply:
I blame copy/paste! (or else I misspelled it "ResportStream" 5 times)